increase max line length in ini files

engelmi opened the pull request:
This PR increases the maximum line length when reading ini files. 
When reading hirte configuration files, the `AllowedNodeNames` entry might get quite long. In order to prevent cutting off the line line reading (and, subsequently, failing) as well as not require to much memory, lets increase the maximum length to a reasonable number. 

Fixes: #425

rhatdan commented on src/libhirte/ini/ini.h:
```diff
@@ -109,7 +109,7 @@ INI_API int ini_parse_string(const char *string, ini_handler handler, void *user
    this must be 3 more than the longest line (due to '\r', '\n', and '\0'). */
 #ifndef INI_MAX_LINE
-#        define INI_MAX_LINE 200
+#        define INI_MAX_LINE 1024
```
This still seems arbitrarily small.  Max DNS name size is 255, if you have 4 nodes each with a QM, that would be 8 node definition + separators would get you to 2040+ chars.

Do we have to have a constant, or could we allocate the buffer on the fly?

engelmi replied:
I agree, its still quite small. In https://github.com/containers/hirte/issues/425#issuecomment-1661693665 I explained why dynamic allocating is not really working out of the box. I'll have a look at it and see if I can write a custom reader function.  

_Update:_ After checking briefly, it seems that implementing a custom reader that wraps `getline` enables us to achive this. It still uses `INI_INITIAL_ALLOC=200` as the initial size, but is able to allocate more memory if needed - and frees it properly (when setting `INI_USE_STACK=0` so the library uses a "dynamic" approach). I reset the constant to its previous value as well. 
Could you have another look? @rhatdan

mwperina commented on src/libhirte/ini/ini.c:
```diff
@@ -257,9 +257,21 @@ int ini_parse_stream(ini_reader reader, void *stream, ini_handler handler, void
+Hirte's custom reader to read a line with arbitrary length.
+This function requires INI_USE_STACK=0.
+*/
+char *dynamic_reader(char *str, int num, void *stream) {
```
We have copied this file from https://github.com/benhoyt/inih so it would be better to leave untouched. Could you please create another C file in the same directory with our customized methods?

engelmi replied:
Sure, I can move it. But using valgrind I also found a memory allocation issue with the solution - the inih library is rather hard to extend in this regards. I'll need to take a deeper look into that and, if I can't fix it, we'll have to fallback to a large enough buffer.

engelmi replied:
Edit: The `ini_parse_file` I need to change nevertheless, since it only allows to pass in a custom handler.

mwperina replied:
> Sure, I can move it. But using valgrind I also found a memory allocation issue with the solution - the inih library is rather hard to extend in this regards. I'll need to take a deeper look into that and, if I can't fix it, we'll have to fallback to a large enough buffer.

The original project https://github.com/benhoyt/inih is active, so could you please retry with the latest and if the issue still exists, could you please report there? And maybe even post a fix if you know how to fix it?

mwperina commented on src/libhirte/ini/ini.h:
```diff
@@ -103,7 +103,7 @@ INI_API int ini_parse_string(const char *string, ini_handler handler, void *user
 /* Nonzero to use stack for line buffer, zero to use heap (malloc/free). */
 #ifndef INI_USE_STACK
-#        define INI_USE_STACK 1
+#        define INI_USE_STACK 0
```
Wouldn't it be better to pass `-DINI_USE_STACK=0` into the project build rather than modifying source code which we copied from different project?

https://github.com/benhoyt/inih/blob/master/README.md#memory-options

engelmi committed:
support multiple lines in config

The inih ini_parse_stream function will call the handler
multiple times when a value in spread across multiple
lines. Previously, this caused the custom handler to
override the value - implementing a last one wins method.
This has been changed to append the value instead. It now
also uses a temporary config for collecting settings from
all files and merging it with the previous one.
An integration test has also been added.

Signed-off-by: Michael Engel <mengel@redhat.com>

mwperina reviewed (approved):
LGTM

mwperina commented on doc/docs/configuration.md:
````diff
@@ -18,3 +20,14 @@ configuration has the highest priority and all defined settings will override pr
+multiple, indented lines. For example, a large number of node names in the `AllowedNodeNames` field can be split like this:
+
+```bash
+AllowedNodeNames=node0,node1,node2,node3,node4,node5,node6,node7,node8,node9,node10,node11,node12,node13,node14,node15,node16,node17,
````
I'd prefer simpler example, following seems to me descriptive enough
```
AllowedNodeNames=node1,
node2,
node3
```

engelmi replied:
Changed

mwperina commented on doc/man/hirte.conf.5.md:
```diff
@@ -64,6 +66,20 @@ LogTarget=journald
+AllowedNodeNames=agent-007,
+   agent-123,
+   agent-456,
+   agent-789,
```
Hmm, wouldn't the `,` after the last node causing parsing error?

engelmi replied:
No. The `,` is independent of the ini parsing. And later on when splitting the names by the `,` symbol we check if the node name is empty or NULL. The `cfg_dump` is a bit misleading (it just takes the read, full string), but via `list_nodes()` API call on the manager this can be verified.

mwperina replied:
OK, that's great. But I'd probably not put the `,` if we don't need it there

engelmi replied:
I agree. Removed it.

mwperina reviewed (approved):
LGTM
